### DISCUSSION TITLE
Final Mini Proejct (Guided) - BlogExplorer - React, Redux, React-Redux, Redux Thunk, Axios, GraphQL (Apollo Client), and mock APIs

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#250] [Tasks]
This is a guided capstone mini-project that incorporates React, Redux, React-Redux, Redux Thunk, Axios, GraphQL (Apollo Client), and the mentioned mock APIs. This project will focus on building a blog application with posts and comments, demonstrating how to query data from both REST and GraphQL APIs.

Project: BlogExplorer

Overview:
BlogExplorer is a React application that fetches blog posts from a REST API and comments from a GraphQL API. Users can view posts, click on a post to see its details and comments, and add new comments.

Step 1: Project Setup

First, create a new React project using Create React App:

```bash
npx create-react-app blog-explorer
cd blog-explorer
```

Install the necessary dependencies:

```bash
npm install redux react-redux redux-thunk axios @apollo/client graphql
```

Step 2: Project Structure

Organize your project structure as follows:

```
src/
  components/
    PostList.js
    PostDetail.js
    CommentList.js
    CommentForm.js
  redux/
    actions/
      postActions.js
      commentActions.js
    reducers/
      postReducer.js
      commentReducer.js
      rootReducer.js
    store.js
  graphql/
    queries.js
  App.js
  index.js
```

Step 3: Setting up Redux

Create the Redux store in `src/redux/store.js`:

```javascript
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './reducers/rootReducer';

const store = createStore(rootReducer, applyMiddleware(thunk));

export default store;
```

Create the root reducer in `src/redux/reducers/rootReducer.js`:

```javascript
import { combineReducers } from 'redux';
import postReducer from './postReducer';
import commentReducer from './commentReducer';

const rootReducer = combineReducers({
  posts: postReducer,
  comments: commentReducer,
});

export default rootReducer;
```

Step 4: Implementing Post Functionality

Create the post reducer in `src/redux/reducers/postReducer.js`:

```javascript
const initialState = {
  posts: [],
  loading: false,
  error: null,
};

const postReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'FETCH_POSTS_REQUEST':
      return { ...state, loading: true };
    case 'FETCH_POSTS_SUCCESS':
      return { ...state, loading: false, posts: action.payload };
    case 'FETCH_POSTS_FAILURE':
      return { ...state, loading: false, error: action.payload };
    default:
      return state;
  }
};

export default postReducer;
```

Create post actions in `src/redux/actions/postActions.js`:

```javascript
import axios from 'axios';

export const fetchPosts = () => async (dispatch) => {
  dispatch({ type: 'FETCH_POSTS_REQUEST' });
  try {
    const response = await axios.get('https://jsonplaceholder.typicode.com/posts');
    dispatch({ type: 'FETCH_POSTS_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({ type: 'FETCH_POSTS_FAILURE', payload: error.message });
  }
};
```

Step 5: Implementing Comment Functionality with GraphQL

Set up Apollo Client in `src/index.js`:

```javascript
import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { ApolloClient, InMemoryCache, ApolloProvider } from '@apollo/client';
import store from './redux/store';
import App from './App';

const client = new ApolloClient({
  uri: 'https://graphqlzero.almansi.me/api',
  cache: new InMemoryCache(),
});

Certainly, I'll continue from where we left off:

```javascript
ReactDOM.render(
  <Provider store={store}>
    <ApolloProvider client={client}>
      <App />
    </ApolloProvider>
  </Provider>,
  document.getElementById('root')
);
```

Create GraphQL queries in `src/graphql/queries.js`:

```javascript
import { gql } from '@apollo/client';

export const GET_COMMENTS = gql`
  query GetComments($postId: ID!) {
    post(id: $postId) {
      comments {
        data {
          id
          name
          email
          body
        }
      }
    }
  }
`;

export const ADD_COMMENT = gql`
  mutation AddComment($input: CreateCommentInput!) {
    createComment(input: $input) {
      id
      name
      email
      body
    }
  }
`;
```

Create the comment reducer in `src/redux/reducers/commentReducer.js`:

```javascript
const initialState = {
  comments: [],
  loading: false,
  error: null,
};

const commentReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'FETCH_COMMENTS_REQUEST':
      return { ...state, loading: true };
    case 'FETCH_COMMENTS_SUCCESS':
      return { ...state, loading: false, comments: action.payload };
    case 'FETCH_COMMENTS_FAILURE':
      return { ...state, loading: false, error: action.payload };
    case 'ADD_COMMENT_SUCCESS':
      return { ...state, comments: [...state.comments, action.payload] };
    default:
      return state;
  }
};

export default commentReducer;
```

Create comment actions in `src/redux/actions/commentActions.js`:

```javascript
import { client } from '../../index';
import { GET_COMMENTS, ADD_COMMENT } from '../../graphql/queries';

export const fetchComments = (postId) => async (dispatch) => {
  dispatch({ type: 'FETCH_COMMENTS_REQUEST' });
  try {
    const { data } = await client.query({
      query: GET_COMMENTS,
      variables: { postId },
    });
    dispatch({ type: 'FETCH_COMMENTS_SUCCESS', payload: data.post.comments.data });
  } catch (error) {
    dispatch({ type: 'FETCH_COMMENTS_FAILURE', payload: error.message });
  }
};

export const addComment = (input) => async (dispatch) => {
  try {
    const { data } = await client.mutate({
      mutation: ADD_COMMENT,
      variables: { input },
    });
    dispatch({ type: 'ADD_COMMENT_SUCCESS', payload: data.createComment });
  } catch (error) {
    console.error('Error adding comment:', error);
  }
};
```

Step 6: Creating React Components

Create `src/components/PostList.js`:

```jsx
import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchPosts } from '../redux/actions/postActions';

const PostList = ({ onSelectPost }) => {
  const dispatch = useDispatch();
  const { posts, loading, error } = useSelector((state) => state.posts);

  useEffect(() => {
    dispatch(fetchPosts());
  }, [dispatch]);

  if (loading) return <div>Loading posts...</div>;
  if (error) return <div>Error: {error}</div>;

  return (
    <div>
      <h2>Blog Posts</h2>
      <ul>
        {posts.map((post) => (
          <li key={post.id} onClick={() => onSelectPost(post.id)}>
            {post.title}
          </li>
        ))}
      </ul>
    </div>
  );
};

export default PostList;
```

Create `src/components/PostDetail.js`:

Let's continue with the `PostDetail.js` component and the remaining components:

```jsx
import React from 'react';
import { useSelector } from 'react-redux';
import CommentList from './CommentList';
import CommentForm from './CommentForm';

const PostDetail = ({ postId }) => {
  const post = useSelector((state) =>
    state.posts.posts.find((p) => p.id === postId)
  );

  if (!post) return <div>No post selected</div>;

  return (
    <div>
      <h2>{post.title}</h2>
      <p>{post.body}</p>
      <CommentList postId={postId} />
      <CommentForm postId={postId} />
    </div>
  );
};

export default PostDetail;
```

Create `src/components/CommentList.js`:

```jsx
import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchComments } from '../redux/actions/commentActions';

const CommentList = ({ postId }) => {
  const dispatch = useDispatch();
  const { comments, loading, error } = useSelector((state) => state.comments);

  useEffect(() => {
    dispatch(fetchComments(postId));
  }, [dispatch, postId]);

  if (loading) return <div>Loading comments...</div>;
  if (error) return <div>Error: {error}</div>;

  return (
    <div>
      <h3>Comments</h3>
      <ul>
        {comments.map((comment) => (
          <li key={comment.id}>
            <strong>{comment.name}</strong> ({comment.email})
            <p>{comment.body}</p>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default CommentList;
```

Create `src/components/CommentForm.js`:

```jsx
import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { addComment } from '../redux/actions/commentActions';

const CommentForm = ({ postId }) => {
  const dispatch = useDispatch();
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [body, setBody] = useState('');

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(addComment({ postId, name, email, body }));
    setName('');
    setEmail('');
    setBody('');
  };

  return (
    <form onSubmit={handleSubmit}>
      <h3>Add a Comment</h3>
      <input
        type="text"
        placeholder="Name"
        value={name}
        onChange={(e) => setName(e.target.value)}
        required
      />
      <input
        type="email"
        placeholder="Email"
        value={email}
        onChange={(e) => setEmail(e.target.value)}
        required
      />
      <textarea
        placeholder="Comment"
        value={body}
        onChange={(e) => setBody(e.target.value)}
        required
      ></textarea>
      <button type="submit">Add Comment</button>
    </form>
  );
};

export default CommentForm;
```

Step 7: Updating App.js

Update `src/App.js` to use the components:

```jsx
import React, { useState } from 'react';
import PostList from './components/PostList';
import PostDetail from './components/PostDetail';

function App() {
  const [selectedPostId, setSelectedPostId] = useState(null);

  return (
    <div className="App">
      <h1>BlogExplorer</h1>
      <div style={{ display: 'flex' }}>
        <div style={{ flex: 1 }}>
          <PostList onSelectPost={setSelectedPostId} />
        </div>
        <div style={{ flex: 2 }}>
          {selectedPostId && <PostDetail postId={selectedPostId} />}
        </div>
      </div>
    </div>
  );
}

export default App;
```

Step 8: Running the Application

Now that we have set up all the components and Redux store, you can run the application:

```bash
npm start
```

This will start the development server, and you should be able to see the BlogExplorer application in your browser.

Final Notes and Best Practices:

1. Code Organization: We've organized the code into logical folders (components, redux, graphql) to maintain a clean project structure.

2. Component Separation: Each component has a single responsibility, making the code more maintainable and easier to understand.

3. Redux Usage: We're using Redux for global state management, with separate actions and reducers for posts and comments.

4. GraphQL Integration: We've integrated Apollo Client for GraphQL queries and mutations, demonstrating how to use it alongside Redux.

5. Error Handling: Basic error handling is implemented in the reducers and components to improve user experience.

6. Async Operations: We're using Redux Thunk for handling asynchronous operations in our action creators.

7. React Hooks: We've used various React hooks (useState, useEffect, useSelector, useDispatch) to manage state and side effects in functional components.

8. Prop Types: For a production application, you should add PropTypes to your components for better type checking.

9. Styling: This example doesn't include CSS. In a real-world application, you'd want to add proper styling, possibly using CSS modules or a CSS-in-JS solution.

10. Testing: For a complete project, you should add unit tests for your components and integration tests for your Redux logic.

11. Performance Optimization: For larger applications, consider implementing performance optimizations like memoization (React.memo, useMemo, useCallback) where appropriate.

This mini-project demonstrates the integration of React, Redux, and GraphQL, providing a practical example of how these technologies can work together in a real-world scenario. It shows how to fetch data from both REST and GraphQL APIs, manage global state with Redux, and update the UI based on user interactions.

To further expand this project, you could add features like pagination for posts and comments, implement a search functionality, or add user authentication. These additions would provide even more opportunities to practice with these technologies and explore more advanced concepts.

**Comment by bedhprasad29** [comment]
Blog Explorer with Authentication, fetching blogs with axios using redux and redux- thunk. And fetching comments using GraphQL.

[https://github.com/bedhprasad29/react-course/tree/master/tasks/blog-explorer](url)

**Comment by vgunaseelan** [comment]
Mini Proejct (BlogExplorer) - using with axios, redux and redux- thunk. And fetching comments using GraphQL.
 
https://github.com/vgunaseelan/blog-explorer

**Comment by santoshbics** [comment]
[final_project.zip](https://github.com/user-attachments/files/16832843/final_project.zip)

**Comment by MrHitss** [comment]
Blog Explorer  - using with axios, redux, redux-thunk and faker. And fetching comments using GraphQL.
https://github.com/MrHitss/blog-ex

**Comment by akashbics** [comment]
Final Mini Project (Guided) - BlogExplorer - React, Redux, React-Redux, Redux Thunk, Axios, GraphQL (Apollo Client), and Mock APIs
[blog-explorer.zip](https://github.com/user-attachments/files/16962841/blog-explorer.zip)

**Comment by nikhilrsuvarna** [comment]
Final Mini Project (Guided) - BlogExplorer - React, Redux, React-Redux, Redux Thunk, Axios, GraphQL (Apollo Client), and Mock APIs
[blog-explorer.zip](https://github.com/user-attachments/files/16977391/blog-explorer.zip)

**Comment by rameshbics** [comment]
Final React Project.
[react-final-assignment.zip](https://github.com/user-attachments/files/16977606/react-final-assignment.zip)

**Comment by FinallyOnline** [comment]
Final Project - https://github.com/FinallyOnline/react-final-assignment/tree/main

**Comment by BharathC** [comment]
Final mini project blog explorer
[final-blog-explorer.zip](https://github.com/user-attachments/files/16989393/final-blog-explorer.zip)

**Comment by nirajkumarb** [comment]
Final mini project blog explorer
[niraj_project.zip](https://github.com/user-attachments/files/16990650/niraj_project.zip)

**Comment by rohitbics** [comment]
Final Project: **Blog Explorer**

[blog-explorer.zip](https://github.com/user-attachments/files/16998772/blog-explorer.zip)
<br />
**Adding Comment:**

![image](https://github.com/user-attachments/assets/06587414-0ba1-4bd1-b369-74fb4156345a)
<br />
**Blog Details:**

![image](https://github.com/user-attachments/assets/819f1c94-6264-4b3d-b7ad-e5c2c8a4113e)

**Reply by rohitkoritech** [reply]
https://github.com/brawnybytes/blog-explorer

**Comment by mohitsen0122** [comment]
[blog-final-project-assesment.zip](https://github.com/user-attachments/files/17023013/blog-final-project-assesment.zip)
